torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -4064,6 +4067,42 @@ Tensor linalg_det_backward(
   }
 }
 
+std::tuple<Tensor, Tensor, Tensor> attn_backward(
+    const Tensor& grad_o,
+    const Tensor& grad_a,
+    const Tensor& q,
+    const Tensor& k,
+    const Tensor& v) {
+  if (!grad_o.defined() && !grad_a.defined()) {
+    return {};
+  }
+
+  Tensor a = at::tanh(at::matmul(q, k.transpose(0, 1)));
+
+  if (grad_o.defined()) {
+    Tensor q_o =
+        at::matmul(at::matmul(grad_o, v.transpose(0, 1)) * (1 - a.pow(2)), k);
+    Tensor k_o = at::matmul(
+        (at::matmul(grad_o, v.transpose(0, 1)) * (1 - a.pow(2)))
+            .transpose(0, 1),
+        q);
+    Tensor v_o = at::matmul(a.transpose(0, 1), grad_o);
+
+    if (!grad_a.defined()) {
+      return std::tuple<Tensor, Tensor, Tensor>(q_o, k_o, v_o);
+    }
+    Tensor q_a = at::matmul(grad_a * (1 - a.pow(2)), k);
```

Comment:
No action needed, but this would be a typical place where one could perform in-place to reduce the amount of memory allocated - here, we could could potentially just start with `q_o` here, and do `q_o.add_(at::matmul(...))`

If you do something like `a.add_(b)` only 2 tensors are in memory at one time, vs `a + b` 3 tensors an in memory.

Reply:
The trickiness though with in-place here, is that subclassing use cases won't work as well, so usually additional logic is needed to handle that.